### DISCUSSION TITLE
Environment variables containing strings with spaces are not accepted any longer for passenv/pass_env.

### DISCUSSION THREAD

**Discussion by Marc2508** [tox-dev/tox#2699] [Q&A] [answered]
Looks like this change:
https://github.com/tox-dev/tox/issues/2658
causes an error for environment variables that are strings:
py3: failed with pass_env values cannot contain whitespace, use comma to have multiple values in a single line, invalid values found '-m "not scale"'

export PYTEST_ARGS = '-m "not scale"'

passenv = 
   {env:PYTEST_ARGS}

Is this intended?

**Comment by mgedmin** [comment, marked as answer]
This seems like weird syntax.  If you want PYTEST_ARGS to be passed to the commands executed by tox, you would put

    pasenv = PYTEST_ARGS

in the tox.ini.  Why are you using `{env:PYTEST_ARGS}`?  {env:...} is for including the value of the environment variable into tox configuration.

**Comment by gaborbernat** [comment]
This seems like a broken configuration where the validation caught it. You're welcome 😊

**Comment by Marc2508** [comment]
Thanks for the fast response and your help! 
The problem was indeed {env:PYTEST_ARGS} instead of just using PYTEST_ARGS. The PYTEST_ARGS are actually exported by an external Makefile so I assume the {env:PYTEST_ARGS} expression was chosen to read this environment variable. This was working in general but it seems to remove the quotes around the expression.

**Comment by mkoeppe** [comment]
My tox.ini files are also affected by the change to the parsing of `passenv` in tox 4: 
I can no longer use factors to conditionalize which environment variables are passed:
```
passenv =
    sagepython, sagewheels:  SAGE_VENV
    sagewheels:              SAGE_SPKG_WHEELS
```
(from https://github.com/sagemath/sage/blob/develop/pkgs/sagemath-categories/tox.ini#L29)

**Reply by gaborbernat** [reply]
This is unrelated to the question here; you should open an issue for it.

**Reply by mkoeppe** [reply]
That's now #2718

**Comment by gaborbernat** [comment]
The original question was resolved. @mkoeppe if you have a separate issue with factors in `passenv` not working, please open a new discussion or issue with a minimal reproducer.

**Reply by mkoeppe** [reply]
#2718 was opened and resolved already
